### DISCUSSION TITLE
Ability to Utilize JDT LS for one-off refactorings via API

### DISCUSSION THREAD

**Discussion by AlexShypula** [redhat-developer/vscode-java#2523] [Q&A]
Hi, 

I'm working on a research project where I'd like to programmatically leverage the JDT LS  extract method functionality. 

As I understand, the language server protocol is a universal protocol between editors and language servers. I was interested in leveraging the fact, and I was curious if there is some way to hack together a way to start up the java language server, and send one-off requests (such as extract a method from this .java file starting at location=XXX and ending at location=XXX). These requests/files/edits would be done entirely programmatically and would have nothing to do with me manually opening VSCode. 

From what I've found out, the JDT LS requires that this VSCode Java server is first started, that way JDT LS can actually start running. If I start the VSCode server so that JDT LS can then begin, is there any way I could try to exploit any of the functionality in the JDT Language Server in the way I'm hoping to?

**Comment by snjeza** [comment]
@AlexShypula You may want to take a look at https://github.com/eclipse/eclipse.jdt.ls/blob/master/org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/refactoring/ExtractMethodTest.java

**Reply by AlexShypula** [reply]
Yeah, I've been looking at this since posting. I was thinking I could interface with the server using json requests, but yes, I think it may be easiest to import parts of the library and utilize those.

**Reply by AlexShypula** [reply]
Thanks for the reply as well : )

**Comment by snjeza** [comment]
@AlexShypula You have to create/import an Eclipse project and java source file. After that you can send and receive json data. ExtractMethodTest is an example.
